hudi-client/hudi-java-client/src/main/java/org/apache/hudi/table/HoodieJavaTable.java

```diff
@@ -93,8 +93,10 @@ protected Option<HoodieTableMetadataWriter> getMetadataWriter(String triggeringI
           getContext().getStorageConf(), config, failedWritesCleaningPolicy, getContext(),
           Option.of(triggeringInstantTimestamp));
       // even with metadata enabled, some index could have been disabled
-      // delete metadata partitions corresponding to such indexes
-      deleteMetadataIndexIfNecessary();
+      // delete metadata partitions corresponding to such indexes if autoDetectAndDeleteMdtPartitions is enabled
+      if (config.isAutoDetectAndDeleteMdtPartitions()) {
+        deleteMetadataIndexIfNecessary();
+      }
```

Comment:
Discussed with @nsivabalan offline that it's better to avoid adding a new internal write config.  Instead, the `getMetadataWriter` API should be revised so that the indexer and upgrade/downgrade path should get the metadata write without calling `deleteMetadataIndexIfNecessary()` based on the method argument, since the relevant write configs may not be available.